improver/wxcode/modal_code.py

```diff
@@ -244,28 +315,456 @@ def process(self, cubes: CubeList) -> Cube:
+                        bounds[0] = bounds[0] + increment
+                        day_slice.coord(coord).bounds = bounds
+                day_cubes.append(day_slice)
+        return day_cubes.concatenate_cube()
+
+    def _find_dry_indices(self, cube: Cube, time_axis: int) -> np.ndarray:
+        """Find the indices indicating dry weather codes. This can include a wet bias
+        if supplied.
+
+        Args:
+            cube: Weather codes cube.
+            time_axis: The time coordinate dimension.
+
+        Returns:
+            Boolean array that is true if the weather codes are dry or False otherwise.
+        """
+        # Find indices corresponding to dry weather codes inclusive of a wet bias.
+        dry_counts = np.sum(
+            np.isin(cube.data, self.broad_categories["dry"]), axis=time_axis
+        )
+        wet_counts = np.sum(
+            np.isin(cube.data, self.broad_categories["wet"]), axis=time_axis
+        )
+        return dry_counts > self.wet_bias * wet_counts
+
+    @staticmethod
+    def counts_per_category(data: np.ndarray) -> np.ndarray:
+        """Implemented following https://stackoverflow.com/questions/46256279/bin-elements-per-row-vectorized-2d-bincount-for-numpy/46256361#46256361
+        Use np.bincount to count the number of occurrences within each category, so that
+        the most common occurrence can then be found.
+
+        Args:
+            data: Array where occurrences of each possible integer value between 0
+                and data.max() will be counted.
+
+        Returns:
+            An array of counts for the occurrence of each category within each row.
+        """
+        n_cat = data.max() + 1
+        a_offs = data + np.arange(data.shape[0])[:, None] * n_cat
+        return np.bincount(a_offs.ravel(), minlength=data.shape[0] * n_cat).reshape(
+            -1, n_cat
+        )
+
+    def _find_most_significant_dry_code(
+        self, cube: Cube, result: Cube, dry_indices: np.ndarray, time_axis: int
+    ) -> Cube:
+        """Find the most significant dry weather code at each point.
+
+        Args:
+            cube: Weather code cube.
+            result: Cube into which to put the result.
+            dry_indices: Boolean, which is true if the weather codes at that point,
+                are dry.
+            time_axis: The time coordinate dimension.
+
+        Returns:
+            Cube where points that are dry are filled with the most common dry
+            code present at that point. If there is a tie, the most significant dry
+            weather code is used, assuming higher values for the weather code indicate
+            more significant weather.
+        """
+        data = cube.data.copy()
+        data = np.ma.masked_where(
+            ~np.isin(cube.data.copy(), self.broad_categories["dry"]), data
+        )
+        counts = self.counts_per_category(data.T)
+        result.data[dry_indices] = np.argmax(counts, axis=1)[dry_indices]
```

Comment:
Should probably use `axis=-1` which would accommodate something odd like a multi-symbol output with a 1st, 2nd, etc. symbol dimension leading the site index. This should no longer be our code by then, but you never know.